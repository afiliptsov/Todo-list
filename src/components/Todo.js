import React, {useState} from 'react'
import Categories from './Categories'
import ListItems from './ListItems'
import defaultState from './state/introState'
import '../styles/ToDo.css'

const Todo = () => {
  const [items, setItems] = useState(defaultState)
  const [userInput, setUserInput] = useState('')

  const addCategory = categoryName => {
    let newCategoryObject = {
      key: categoryName,
      active: false,
      id: Math.floor(Math.random() * 10000),
      list: [],
    }
    setItems([...items, newCategoryObject])
    // Cleaning input field
    setUserInput('')
  }

  const addListItem = listItem => {
    let findActiveCategory = items.filter(item => {
      return item.active
    })
    if (findActiveCategory.length !== 1) {
      alert('Please select active Category. Only 1 allowed.')
    } else {
      let listItemObject = {
        id: Math.floor(Math.random() * 10000),
        description: listItem,
        checked: false,
      }
      let addListItem = items.map(item =>
        item.active === true
          ? {...item, list: [...item.list, listItemObject]}
          : item,
      )
      setItems(addListItem)
      // Cleaning input field
      setUserInput('')
    }
  }

  const handleFormSubmit = (event, section) => {
    event.preventDefault()
    if (userInput.length === 0) {
      alert('Your input should have at least 1 character')
    } else {
      if (section === 'category') {
        addCategory(userInput)
      } else if (section === 'listItem') {
        addListItem(userInput)
      }
    }
  }

  const removeCategory = categoryId => {
    let updatedCategories = items.filter(item =>
      item.id === categoryId ? item.id !== categoryId : item,
    )
    setItems(updatedCategories)
  }

  const removeTodoItem = (categoryId, listItemId) => {
    let updatedItems = items.map(item =>
      item.id === categoryId
        ? {...item, list: item.list.filter(x => x.id !== listItemId.id)}
        : item,
    )
    setItems(updatedItems)
  }

  const toggleItem = (categoryId, listItemId) => {
    // This covers Check and UnCheck functionality for items list
    let checkedItem = items.map(item =>
      item.id === categoryId
        ? {
            ...item,
            list: item.list.map(x => {
              if (x.id === listItemId.id) {
                x.checked = !x.checked
              }
              return x
            }),
          }
        : item,
    )
    setItems(checkedItem)
  }

  const toggleCategory = categoryId => {
    let testItem = items.map(item => {
      if (item.id === categoryId) {
        item.active = !item.active
      }
      return item
    })
    setItems(testItem)
  }

  return (
    <div className="todoWrapper">
      <div className="todoHeader">
        <h2 data-testid="todoListHeader">My ToDo List</h2>
        <label>{new Date().toDateString()}</label>
      </div>
      <div className="todoContent">
        <Categories
          data-testid="categories"
          toggleCategory={toggleCategory}
          removeCategory={removeCategory}
          categoriesList={items}
        />
        <ListItems
          data-testid="listItems"
          toggleItem={toggleItem}
          removeTodoItem={removeTodoItem}
          todoListItems={items.filter(category => category.active)}
        />
      </div>
      <form
        data-testid="submitForm"
        className="addSection"
        onSubmit={handleFormSubmit}
      >
        <input
          data-testid="textInput"
          type="text"
          value={userInput}
          onChange={e => setUserInput(e.target.value)}
          placeholder="Category or Item"
        />
        <button
          data-testid="categorySubmit"
          id="category_submit"
          onClick={e => handleFormSubmit(e, 'category')}
        >
          Add Category
        </button>
        <button
          data-testid="listItemSubmit"
          id="listItem_submit"
          onClick={e => handleFormSubmit(e, 'listItem')}
        >
          Add item
        </button>
      </form>
    </div>
  )
}

export default Todo
