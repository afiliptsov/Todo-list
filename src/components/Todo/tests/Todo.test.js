import {fireEvent, render, screen} from '@testing-library/react'
import renderer from 'react-test-renderer'
import mockState from '../../../state/mockState'
import Todo from '../../Todo/Todo'

describe('Todo page elements are displayed', () => {
  it('should render Todo page Header', async () => {
    render(<Todo />)
    const getTitle = screen.getByTestId('todoListHeader')
    expect(getTitle).toBeVisible()
  })
  it('should display Add Category Submit button', async () => {
    render(<Todo />)
    const getCategorySubmitButton = screen.getByTestId('categorySubmit')
    expect(getCategorySubmitButton).toHaveTextContent('Add Category')
  })
  it('should display Add List Item Submit button', async () => {
    render(<Todo />)
    const getListItemSubmitButton = screen.getByTestId('listItemSubmit')
    expect(getListItemSubmitButton).toHaveTextContent('Add item')
  })
  it('should display user input field', async () => {
    render(<Todo />)
    const getTextInput = screen.getByTestId('textInput')
    expect(getTextInput).toBeVisible()
  })
})

describe('Input', () => {
  it('should be able to type input', async () => {
    render(<Todo />)
    const inputElement = screen.getByTestId('textInput')
    fireEvent.change(inputElement, {target: {value: 'sports'}})
    expect(inputElement.value).toBe('sports')
  })

  it('input should be cleared after category submission', async () => {
    const testCategory = 'sport'
    render(<Todo />)
    const inputElement = screen.getByTestId('textInput')
    const categoryButton = screen.getByTestId('categorySubmit')
    fireEvent.change(inputElement, {target: {value: testCategory}})
    expect(inputElement.value).toBe(testCategory)
    fireEvent.click(categoryButton)
    expect(inputElement.value).toBe('')
  })
})

describe('Todo snapshot validation', () => {
  it('Todo should match snapshot', () => {
    const component = renderer.create(<Todo />).toJSON()
    expect(component).toMatchSnapshot()
  })
})
