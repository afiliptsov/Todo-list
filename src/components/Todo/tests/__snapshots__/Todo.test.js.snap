// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo snapshot validation Todo should match snapshot 1`] = `
<div
  className="todoWrapper"
>
  <div
    className="todoHeader"
  >
    <h2
      data-testid="todoListHeader"
    >
      My ToDo List
    </h2>
    <label>
      Thu Jun 16 2022
    </label>
  </div>
  <div
    className="todoContent"
  >
    <div
      className="categoriesWrapper"
      data-testid="categoriesComponent"
    >
      <ul>
        <p
          className="categorySection"
        >
          <li
            className="listCategory activeCategory"
            onClick={[Function]}
          >
            <span
              data-testid="intro"
            >
              intro
            </span>
          </li>
          <svg
            className="icon"
            data-testid="intro-delete"
            fill="none"
            height="1em"
            onClick={[Function]}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 4H8l-7 8 7 8h13a2 2 0 0 0 2-2V6a2 2 0 0 0-2-2z"
            />
            <line
              x1="18"
              x2="12"
              y1="9"
              y2="15"
            />
            <line
              x1="12"
              x2="18"
              y1="9"
              y2="15"
            />
          </svg>
        </p>
      </ul>
    </div>
    <div
      className="toDoListWrapper"
    >
      <ul>
        <div
          className="toDoListItem"
        >
          <div
            className="todoListItemIcon"
            onClick={[Function]}
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
              />
            </svg>
          </div>
          <div>
            <li>
              <div
                className="categoryToggle"
                data-testid="Hi there "
              >
                Hi there 
              </div>
            </li>
          </div>
          <div
            className="todoListItemIcon todoListItemDeleteIcon"
          >
            <svg
              baseProfile="tiny"
              fill="currentColor"
              height="1em"
              onClick={[Function]}
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              version="1.2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 3c-4.963 0-9 4.038-9 9s4.037 9 9 9 9-4.038 9-9-4.037-9-9-9zm0 16c-3.859 0-7-3.14-7-7s3.141-7 7-7 7 3.14 7 7-3.141 7-7 7zM12.707 12l2.646-2.646c.194-.194.194-.512 0-.707-.195-.194-.513-.194-.707 0l-2.646 2.646-2.646-2.647c-.195-.194-.513-.194-.707 0-.195.195-.195.513 0 .707l2.646 2.647-2.646 2.646c-.195.195-.195.513 0 .707.097.098.225.147.353.147s.256-.049.354-.146l2.646-2.647 2.646 2.646c.098.098.226.147.354.147s.256-.049.354-.146c.194-.194.194-.512 0-.707l-2.647-2.647z"
              />
            </svg>
          </div>
        </div>
        <div
          className="toDoListItem"
        >
          <div
            className="todoListItemIcon"
            onClick={[Function]}
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
              />
              <path
                d="M10.996 12.556 9.7 11.285l-1.4 1.43 2.704 2.647 4.699-4.651-1.406-1.422z"
              />
            </svg>
          </div>
          <div>
            <li>
              <div
                className="categoryToggle"
                data-testid="<--- You can check items by clicking checkbox"
              >
                &lt;--- You can check items by clicking checkbox
              </div>
            </li>
          </div>
          <div
            className="todoListItemIcon todoListItemDeleteIcon"
          >
            <svg
              baseProfile="tiny"
              fill="currentColor"
              height="1em"
              onClick={[Function]}
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              version="1.2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 3c-4.963 0-9 4.038-9 9s4.037 9 9 9 9-4.038 9-9-4.037-9-9-9zm0 16c-3.859 0-7-3.14-7-7s3.141-7 7-7 7 3.14 7 7-3.141 7-7 7zM12.707 12l2.646-2.646c.194-.194.194-.512 0-.707-.195-.194-.513-.194-.707 0l-2.646 2.646-2.646-2.647c-.195-.194-.513-.194-.707 0-.195.195-.195.513 0 .707l2.646 2.647-2.646 2.646c-.195.195-.195.513 0 .707.097.098.225.147.353.147s.256-.049.354-.146l2.646-2.647 2.646 2.646c.098.098.226.147.354.147s.256-.049.354-.146c.194-.194.194-.512 0-.707l-2.647-2.647z"
              />
            </svg>
          </div>
        </div>
        <div
          className="toDoListItem"
        >
          <div
            className="todoListItemIcon"
            onClick={[Function]}
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
              />
            </svg>
          </div>
          <div>
            <li>
              <div
                className="categoryToggle"
                data-testid="You can delete items too --->"
              >
                You can delete items too ---&gt;
              </div>
            </li>
          </div>
          <div
            className="todoListItemIcon todoListItemDeleteIcon"
          >
            <svg
              baseProfile="tiny"
              fill="currentColor"
              height="1em"
              onClick={[Function]}
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              version="1.2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 3c-4.963 0-9 4.038-9 9s4.037 9 9 9 9-4.038 9-9-4.037-9-9-9zm0 16c-3.859 0-7-3.14-7-7s3.141-7 7-7 7 3.14 7 7-3.141 7-7 7zM12.707 12l2.646-2.646c.194-.194.194-.512 0-.707-.195-.194-.513-.194-.707 0l-2.646 2.646-2.646-2.647c-.195-.194-.513-.194-.707 0-.195.195-.195.513 0 .707l2.646 2.647-2.646 2.646c-.195.195-.195.513 0 .707.097.098.225.147.353.147s.256-.049.354-.146l2.646-2.647 2.646 2.646c.098.098.226.147.354.147s.256-.049.354-.146c.194-.194.194-.512 0-.707l-2.647-2.647z"
              />
            </svg>
          </div>
        </div>
        <div
          className="toDoListItem"
        >
          <div
            className="todoListItemIcon"
            onClick={[Function]}
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
              />
            </svg>
          </div>
          <div>
            <li>
              <div
                className="categoryToggle"
                data-testid="You can display multiple categories at the same time"
              >
                You can display multiple categories at the same time
              </div>
            </li>
          </div>
          <div
            className="todoListItemIcon todoListItemDeleteIcon"
          >
            <svg
              baseProfile="tiny"
              fill="currentColor"
              height="1em"
              onClick={[Function]}
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              version="1.2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 3c-4.963 0-9 4.038-9 9s4.037 9 9 9 9-4.038 9-9-4.037-9-9-9zm0 16c-3.859 0-7-3.14-7-7s3.141-7 7-7 7 3.14 7 7-3.141 7-7 7zM12.707 12l2.646-2.646c.194-.194.194-.512 0-.707-.195-.194-.513-.194-.707 0l-2.646 2.646-2.646-2.647c-.195-.194-.513-.194-.707 0-.195.195-.195.513 0 .707l2.646 2.647-2.646 2.646c-.195.195-.195.513 0 .707.097.098.225.147.353.147s.256-.049.354-.146l2.646-2.647 2.646 2.646c.098.098.226.147.354.147s.256-.049.354-.146c.194-.194.194-.512 0-.707l-2.647-2.647z"
              />
            </svg>
          </div>
        </div>
      </ul>
    </div>
  </div>
  <form
    className="addSection"
    data-testid="submitForm"
    onSubmit={[Function]}
  >
    <input
      data-testid="textInput"
      onChange={[Function]}
      placeholder="Category or Item"
      type="text"
      value=""
    />
    <button
      data-testid="categorySubmit"
      id="category_submit"
      onClick={[Function]}
    >
      Add Category
    </button>
    <button
      data-testid="listItemSubmit"
      id="listItem_submit"
      onClick={[Function]}
    >
      Add item
    </button>
  </form>
</div>
`;
